walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,78 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    int duration = (int) request.getDuration();
+    Set<String> attendees = new HashSet<String>(request.getAttendees());
+    List<TimeRange> initalRanges = new ArrayList<TimeRange>();
+    List<TimeRange> results = new ArrayList<TimeRange>();
+    final int minMeetingTime = 15;
+    
+    for(int start = 0; start <= TimeRange.END_OF_DAY; start += minMeetingTime) {
+        if(start + minMeetingTime > TimeRange.END_OF_DAY) {
+            initalRanges.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, false));
+        } else {
+            initalRanges.add(TimeRange.fromStartDuration(start, minMeetingTime));
+        }
+    }
+
+    for(Event currEvent : events) {
+        Set<String> currAttendees = currEvent.getAttendees();
+        for(String attendee : currAttendees) {
+            if(attendees.contains(attendee)) {
+                blockTimeRange(initalRanges, currEvent);
+                break;
+            }
+        }
+    }
+
+    int newRangeStart = -1;
+    int newRangeDuration = 0;
```

Comment:
This represents the next free block of time? What if you used a TimeRange to represent it instead? You could use null  to represent an absence of a free block instead of an -1 sentinel.

Rationale: this would push the design to be more encapsulated. We could eventually push even more logic into the range class itself, such as Overlaps(), MergeWith(Range other), etc.

Reply:
done!